In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
from bs4 import BeautifulSoup
from requests import get
import urllib3

In [3]:
url1 = "https://www.flipkart.com/search?q=vivo&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off"

connect = get(url1)

connect

<Response [200]>

In [4]:
bs4_obj = BeautifulSoup(connect.text, 'html.parser')

In [5]:
# one full phone container detail
master_containers = bs4_obj.find_all('div', class_ = "_2kHMtA")

len(master_containers)

24

In [6]:
# name of phone
master_containers[0].find('div', class_ = "_4rR01T").text

'vivo T2x 5G (Black Gladiator, 128 GB)'

In [7]:
# flipkart price
master_containers[0].find('div', class_ = "_30jeq3 _1_WHN1").text

'₹14,999'

In [8]:
# rating
master_containers[0].find('div', class_ = "_3LWZlK").text

'4.3'

In [9]:
# actual price
master_containers[0].find('div', class_ = "_3I9_wc _27UcVY").text

'₹20,999'

In [10]:
# Lets scrap the data from first 8 pages of flipkart url for vivo phones

# Lets create the empty list

phone_name = []
flipkart_price = []
rating = []
actual_price = []

# First for loop to create dynamic URL

for x in range (1,9):
    
    # STEP 1: Create URL
    
    url = "https://www.flipkart.com/search?q=vivo&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off"+str(x)
    
    # STEP 2: To request the server to connect to this url

    connect = get(url)
    
    # STEP 3: From the above connection, we will create a bs4 object which contain only text from this url page

    bs4_obj = BeautifulSoup(connect.text, 'html.parser')
    
    # STEP 4: Using this bs4_obj lets get the access to our container

    master_containers = bs4_obj.find_all('div', class_ = "_2kHMtA")
    
    # Writing second for loop to itrate in master containers of each page
    
    for y in master_containers:
        
        # Scraping the model name of all products from first 8 pages
    
        pro_name = y.find('div', class_ = "_4rR01T").text
    
        # Appending the model name list
    
        phone_name.append(pro_name)
    
        # Scraping the flipkart price of all products from first 8 pages
    
        pro_price = y.find('div', class_ = "_30jeq3 _1_WHN1").text
    
        # appending flipkart price list
    
        flipkart_price.append(pro_price)
        
        # Scraping the rating: As rating have missing value, we are replacing 
        # them by NA
    
        if y.find('div', class_ = "_3LWZlK") is not None:
        
            pro_rating = y.find('div', class_ = "_3LWZlK").text
    
        else:
        
            pro_rating = "NA"
    
        # Appending the rating list
    
        rating.append(pro_rating)
        
        # Scraping the actual price of all products from first 8 pages
    
        act_price = y.find('div', class_ = '_3I9_wc _27UcVY').text
        
        # appending actual price list
        actual_price.append(act_price) 

In [11]:
df = pd.DataFrame({"Model Name": phone_name,"Flipkart price": flipkart_price,"Rating" : rating,"Actual price" : actual_price})
df

Model Name Flipkart price Rating  \
0      vivo T2x 5G (Black Gladiator, 128 GB)        ₹14,999    4.3   
1          vivo T2x 5G (Marine Blue, 128 GB)        ₹12,999    4.4   
2      vivo T2x 5G (Black Gladiator, 128 GB)        ₹12,999    4.4   
3      vivo T2x 5G (Sunstone Orange, 128 GB)        ₹12,999    4.4   
4        vivo T2x 5G (Glimmer Black, 128 GB)        ₹12,999    4.4   
..                                       ...            ...    ...   
187  vivo T2 Pro 5G (New Moon Black, 128 GB)        ₹23,999    4.5   
188    vivo T2x 5G (Black Gladiator, 128 GB)        ₹13,999    4.4   
189    vivo T2x 5G (Sunstone Orange, 128 GB)        ₹13,999    4.4   
190       vivo V30 5G (Andaman Blue, 256 GB)        ₹37,999     NA   
191      vivo V30 5G (Peacock Green, 256 GB)        ₹37,999     NA   

    Actual price  
0        ₹20,999  
1        ₹18,999  
2        ₹18,999  
3        ₹18,999  
4        ₹18,999  
..           ...  
187      ₹26,999  
188      ₹17,999  
189      ₹17,999  
190      ₹42,999  
191      ₹42,999  

[192 rows x 4 columns]

In [12]:
df.dtypes

Model Name        object
Flipkart price    object
Rating            object
Actual price      object
dtype: object

In [14]:
df.isnull().sum()

Model Name        0
Flipkart price    0
Rating            0
Actual price      0
dtype: int64

In [15]:
df[df['Model Name']=='NA']

Empty DataFrame
Columns: [Model Name, Flipkart price, Rating, Actual price]
Index: []

In [16]:
df[df['Flipkart price']=='NA']

Empty DataFrame
Columns: [Model Name, Flipkart price, Rating, Actual price]
Index: []

In [17]:
df[df['Actual price']=='NA']

Empty DataFrame
Columns: [Model Name, Flipkart price, Rating, Actual price]
Index: []

In [18]:
df[df['Rating']=='NA']

Model Name  Flipkart price Rating  \
22       vivo V30 5G (Andaman Blue, 256 GB)         37999.0     NA   
23      vivo V30 5G (Peacock Green, 256 GB)         37999.0     NA   
43      vivo V30 5G (Peacock Green, 128 GB)         33999.0     NA   
44   vivo V30 Pro 5G (Andaman Blue, 256 GB)         41999.0     NA   
70       vivo V30 5G (Andaman Blue, 256 GB)         37999.0     NA   
71      vivo V30 5G (Peacock Green, 256 GB)         37999.0     NA   
94      vivo V30 5G (Peacock Green, 128 GB)         33999.0     NA   
95   vivo V30 Pro 5G (Andaman Blue, 256 GB)         41999.0     NA   
118     vivo V30 5G (Peacock Green, 128 GB)         33999.0     NA   
119  vivo V30 Pro 5G (Andaman Blue, 256 GB)         41999.0     NA   
142      vivo V30 5G (Andaman Blue, 256 GB)         37999.0     NA   
143     vivo V30 5G (Peacock Green, 256 GB)         37999.0     NA   
166     vivo V30 5G (Peacock Green, 128 GB)         33999.0     NA   
167  vivo V30 Pro 5G (Andaman Blue, 256 GB)         41999.0     NA   
190      vivo V30 5G (Andaman Blue, 256 GB)         37999.0     NA   
191     vivo V30 5G (Peacock Green, 256 GB)         37999.0     NA   

     Actual price  
22        42999.0  
23        42999.0  
43        38999.0  
44        46999.0  
70        42999.0  
71        42999.0  
94        38999.0  
95        46999.0  
118       38999.0  
119       46999.0  
142       42999.0  
143       42999.0  
166       38999.0  
167       46999.0  
190       42999.0  
191       42999.0

In [19]:
df['Rating'].value_counts()

4.4    93
4.3    50
4.5    32
NA     16
4.2     1
Name: Rating, dtype: int64

In [20]:
# Replacing string NA with NaN or null values
df['Rating'].replace('NA', pd.NA, inplace=True)

In [21]:
# Drop rows with NaN values in the 'Rating' column
df = df.dropna(subset=['Rating'])

In [22]:
# # Convert categorical data from columns to numerical

df['Flipkart price'] = df['Flipkart price'].str.replace('₹', '').str.replace(',', '').astype(float)
df['Actual price'] = df['Actual price'].str.replace('₹', '').str.replace(',', '').astype(float)
df['Rating'] = df['Rating'].astype(float)

# Which model have highest discount?

--> vivo T2 5G

In [23]:
df['Discount']=df['Actual price'] - df['Flipkart price']

In [24]:
df

Model Name  Flipkart price  Rating  \
0      vivo T2x 5G (Black Gladiator, 128 GB)         14999.0     4.3   
1          vivo T2x 5G (Marine Blue, 128 GB)         12999.0     4.4   
2      vivo T2x 5G (Black Gladiator, 128 GB)         12999.0     4.4   
3      vivo T2x 5G (Sunstone Orange, 128 GB)         12999.0     4.4   
4        vivo T2x 5G (Glimmer Black, 128 GB)         12999.0     4.4   
..                                       ...             ...     ...   
185  vivo T2 Pro 5G (New Moon Black, 256 GB)         24999.0     4.5   
186       vivo T2 Pro 5G (Dune Gold, 256 GB)         24999.0     4.5   
187  vivo T2 Pro 5G (New Moon Black, 128 GB)         23999.0     4.5   
188    vivo T2x 5G (Black Gladiator, 128 GB)         13999.0     4.4   
189    vivo T2x 5G (Sunstone Orange, 128 GB)         13999.0     4.4   

     Actual price  Discount  
0         20999.0    6000.0  
1         18999.0    6000.0  
2         18999.0    6000.0  
3         18999.0    6000.0  
4         18999.0    6000.0  
..            ...       ...  
185       27999.0    3000.0  
186       27999.0    3000.0  
187       26999.0    3000.0  
188       17999.0    4000.0  
189       17999.0    4000.0  

[176 rows x 5 columns]

In [25]:
df.loc[df['Discount'].idxmax()]

Model Name        vivo T2 5G (Velocity Wave, 128 GB)
Flipkart price                               15999.0
Rating                                           4.4
Actual price                                 23999.0
Discount                                      8000.0
Name: 13, dtype: object

# What is the average price of vivo mobiles with 10000 INR and above?

--> Actual price = 21629.54;
    Flipkart price = 16207.09

In [26]:
df[df['Actual price']>=10000]['Actual price'].mean()

21629.545454545456

In [27]:
df[df['Actual price']>=10000].mean()

Flipkart price    16097.085227
Rating                4.388636
Actual price      21629.545455
Discount           5532.460227
dtype: float64

In [28]:
df[df['Flipkart price'] >= 10000]['Flipkart price'].mean()

16207.092485549132

In [29]:
df[df['Flipkart price'] >= 10000].mean()

Flipkart price    16207.092486
Rating                4.390751
Actual price      21738.884393
Discount           5531.791908
dtype: float64

# What is the average rating for vivo mobile of below 10000 INR and above 10000 INR?

--> Actual price ( > = 4.38, < no data); 
    Flipkart price ( > =4.39, < 4.26)

In [30]:
df[(df['Actual price'] > 10000)].mean()

Flipkart price    16097.085227
Rating                4.388636
Actual price      21629.545455
Discount           5532.460227
dtype: float64

In [31]:
df[(df['Actual price'] < 10000)].mean()

Model Name        NaN
Flipkart price    NaN
Rating            NaN
Actual price      NaN
Discount          NaN
dtype: object

In [32]:
df[(df['Flipkart price'] > 10000)].mean()

Flipkart price    16207.092486
Rating                4.390751
Actual price      21738.884393
Discount           5531.791908
dtype: float64

In [33]:
df[(df['Flipkart price'] < 10000)].mean()

Flipkart price     9753.333333
Rating                4.266667
Actual price      15324.333333
Discount           5571.000000
dtype: float64

In [35]:
df.to_csv('vivo.csv',index=False)